### PR TITLE
Dev/xinclude

### LINKED ISSUE #415: xinclude in XmlDom extension not implemented
PHPUnit uses `xinclude` in the XmlDom extension. Exception details below:

```
Unhandled Exception: System.NotImplementedException: The method or operation is not implemented.
   at Peachpie.Library.XmlDom.DOMDocument.xinclude(Int32 options)
   at PHPUnit.Util.Xml.load(Context <ctx>, PhpValue actual, PhpValue isHtml, PhpValue filename, PhpValue xinclude, PhpValue strict) in C:\Development\laravel-peachpie-sample\Laravel\vendor\phpunit\phpunit\src\Util\Xml.php:line 87
   at PHPUnit.Util.Xml.loadFile(Context <ctx>, PhpValue filename, PhpValue isHtml, PhpValue xinclude, PhpValue strict) in C:\Development\laravel-peachpie-sample\Laravel\vendor\phpunit\phpunit\src\Util\Xml.php:line 143
   at PHPUnit.Util.Configuration.__construct(PhpValue filename) in C:\Development\laravel-peachpie-sample\Laravel\vendor\phpunit\phpunit\src\Util\Configuration.php:line 84
   at PHPUnit.Util.Configuration..ctor(Context <ctx>, PhpValue filename)
   at PHPUnit.Util.Configuration.getInstance(Context <ctx>, PhpValue filename) in C:\Development\laravel-peachpie-sample\Laravel\vendor\phpunit\phpunit\src\Util\Configuration.php:line 70
   at PHPUnit.TextUI.Command.handleArguments(PhpValue argv) in C:\Development\laravel-peachpie-sample\Laravel\vendor\phpunit\phpunit\src\TextUI\Command.php:line 847
   at PHPUnit.TextUI.Command.run(PhpValue argv, PhpValue exit) in C:\Development\laravel-peachpie-sample\Laravel\vendor\phpunit\phpunit\src\TextUI\Command.php:line 174
   at PHPUnit.TextUI.Command.main(Context <ctx>, PhpTypeInfo <static>, PhpValue exit) in C:\Development\laravel-peachpie-sample\Laravel\vendor\phpunit\phpunit\src\TextUI\Command.php:line 163
   at <Root>.main_php.main(Context <ctx>) in C:\Development\laravel-peachpie-sample\Laravel\main.php:line 54
   at <Root>.main_php.<Main>(Context <ctx>, PhpArray <locals>, Object this, RuntimeTypeHandle <self>) in C:\Development\laravel-peachpie-sample\Laravel\main.php:line 58
   at <Script>.Main(String[] args)
```

### PR DESCRIPTION
**Pull request by TomatorCZ**
Fixes #415

XInclude implementation

### CODE REVIEW

**Comment by jakubmisek** [comment]
Thanks! there is one more compatibility issue with the PR, it seams on Windows the extension uses `\n` instead of `\r\n`
```
Expected: ···?xml version="1.0"?>\n<root xmlns:xi="http://www.w3.org/2001/X···
Actual:   ···?xml version="1.0"?>\r\n<root xmlns:xi="http://www.w3.org/2001/···
```

any idea why it changes or substitutes the line endings?

**Comment by TomatorCZ** [comment]
> Thanks! there is one more compatibility issue with the PR, it seams on Windows the extension uses `\n` instead of `\r\n`
> 
> ```
> Expected: ···?xml version="1.0"?>\n<root xmlns:xi="http://www.w3.org/2001/X···
> Actual:   ···?xml version="1.0"?>\r\n<root xmlns:xi="http://www.w3.org/2001/···
> ```
> 
> any idea why it changes or substitutes the line endings?

Problem is in method Load of XmlDocument. When you use Load(string path) then new line is \r\n. But peachpie uses Load(XmlReader reader) where new line is represented by \n after load.

**Comment by jakubmisek** [comment]
@TomatorCZ don't think so,  `XmlWriterSettings` has property [NewLineHandling](https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmlwritersettings.newlinehandling?view=netframework-4.8) which defaults to `Replace` and causing the new lines to be replaced with [NewLineChars](https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmlwritersettings.newlinechars?view=netframework-4.8). We should probably set `NewLineHandling` to `None` at https://github.com/peachpiecompiler/peachpie/blob/020eb0d5290c4bd9e6676e9c9b2384ba02023d1b/src/Peachpie.Library.XmlDom/DOMDocument.cs#L764 ?

**Comment by jakubmisek** [comment]
> > Replace is the default value and this change has no effect.
> 
> Yes, I know sorry. I updated it

please run tests locally before commiting! :)

**Comment by TomatorCZ** [comment]
Setting NewLineHandeling solves problems with new line characters. But Php adds another new line after root node. So I saw hexviewer for checking endline after root node. There were no newline chars. When I tried to add new line to file, everything was ok.

**Comment by TomatorCZ** [comment]
Should I repair tests or code ?

**Comment by diddledan** [comment]
Might be worth adding "Fixes #415" to the first post in this thread to autoclose the issue when it is merged :-)

File: src/Tests/Peachpie.Test/Peachpie.Test.csproj [hunk 1]
```diff
@@ -11,6 +11,7 @@
     <ProjectReference Include="..\..\Peachpie.Library.Graphics\Peachpie.Library.Graphics.csproj" />
     <ProjectReference Include="..\..\Peachpie.Library.Network\Peachpie.Library.Network.csproj" />
     <ProjectReference Include="..\..\Peachpie.Library.Scripting\Peachpie.Library.Scripting.csproj" />
+    <ProjectReference Include="..\..\Peachpie.Library.XmlDom\Peachpie.Library.XmlDom.csproj" />
```
Comment (jakubmisek) [inline src/Tests/Peachpie.Test/Peachpie.Test.csproj]: not relevant to this PR but we can keep it - just be careful next time when committing something

File: src/Peachpie.Library.XmlDom/XIncludeHelper.cs [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+                Fixlang(parent, document);
+
+                XmlNode importedNode = parent.ImportNode(document.DocumentElement, true); //Import method changes baseuri instead of does not change
+                includeNode.ParentNode.ReplaceChild(importedNode, includeNode);
+                replaceCount++;
+
+                return;
+            }
+            else
+                replaceCount = -1;
```
Comment (jakubmisek) [inline src/Peachpie.Library.XmlDom/XIncludeHelper.cs]: probably misusing a variable for something else than "count" :)

File: src/Peachpie.Library.XmlDom/XIncludeHelper.cs [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+                    lang.Value = child.DocumentElement.Attributes.GetNamedItem(atLang).Value;
+
+                    child.DocumentElement.Attributes.Append(lang);
+                }
+        }
+
+        void IncludeText(string absoluteUri, XmlElement includeNode)
+        {
+            // Reading xml as a text
+            StreamReader sr = new StreamReader(absoluteUri);
```
Comment (jakubmisek) [inline src/Peachpie.Library.XmlDom/XIncludeHelper.cs]: use the `using` pattern instead
use `var`

File: src/Peachpie.Library.XmlDom/XIncludeHelper.cs [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+        public const string atXPointer = "xpointer";
+        public const string atBase = "base";
+        public const string atLang = "lang";
+        public const string atEncoding = "encoding";
+        #endregion
+        
+        Stack<XmlDocument> documents = new Stack<XmlDocument>(); // When IncludeXml detects tag include, pushes curent document to documents stack and resolves next documents, after that, pulls previous document in other to includes probing document.
+        Dictionary<string, string> references = new Dictionary<string, string>(); // Keeps route of calling documents and controls loop
+        int replaceCount = 0; // Counts include replaces.
+        Context ctx;
```
Comment (jakubmisek) [inline src/Peachpie.Library.XmlDom/XIncludeHelper.cs]: `readonly`

File: src/Peachpie.Library.XmlDom/XIncludeHelper.cs [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        // Atributes
+        public const string atHref = "href";
+        public const string atParse = "parse";
+        public const string atXPointer = "xpointer";
+        public const string atBase = "base";
+        public const string atLang = "lang";
+        public const string atEncoding = "encoding";
+        #endregion
+        
+        Stack<XmlDocument> documents = new Stack<XmlDocument>(); // When IncludeXml detects tag include, pushes curent document to documents stack and resolves next documents, after that, pulls previous document in other to includes probing document.
```
Comment (jakubmisek) [inline src/Peachpie.Library.XmlDom/XIncludeHelper.cs]: unnecessary object allocation, you do it again in `Include`

File: src/Peachpie.Library.XmlDom/XIncludeHelper.cs [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+        Dictionary<string, string> references = new Dictionary<string, string>(); // Keeps route of calling documents and controls loop
```
Comment (jakubmisek) [inline src/Peachpie.Library.XmlDom/XIncludeHelper.cs]: unnecessary object allocation, you do it again in `Include`

File: src/Peachpie.Library.XmlDom/XIncludeHelper.cs [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+                        PhpLibXml.IssueXmlError(ctx, PhpLibXml.LIBXML_ERR_WARNING, 0, 0, 0, $"DOMDocument::xinclude(): detected a local recursion with no xpointer in {absoluteUri} in {ctx.MainScriptFile.Path}", absoluteUri);
+                        return true;
+                    }
+                    documents.Push(document);
+                    if (references.ContainsKey(includeElement.BaseURI + valueOfXPoiner)) //fatal error, cycle recursion
+                    {
+                        PhpLibXml.IssueXmlError(ctx, PhpLibXml.LIBXML_ERR_WARNING,0,0,0, $"DOMDocument::xinclude(): detected a recursion in {absoluteUri} in {ctx.MainScriptFile.Path}", absoluteUri);
+                        return true; 
+                    }
+                    references[absoluteUri] = includeElement.BaseURI + valueOfXPoiner;
```
Comment (jakubmisek) [inline src/Peachpie.Library.XmlDom/XIncludeHelper.cs]: duplicit code `includeElement.BaseURI + valueOfXPoiner`

File: src/Peachpie.Library.XmlDom/XIncludeHelper.cs [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+        /// <param name="nsm">Namespace Manager of document</param>
+        /// <returns></returns>
+        bool TreatIncludes(XmlElement[] includeNodes, XmlDocument document, string absoluteUri, string nsPrefix, XmlNamespaceManager nsm)
+        {
+            foreach (XmlElement includeElement in includeNodes)
+            {
+                string valueOfXPoiner = includeElement.GetAttribute(atXPointer);
+                string valueOfHref = includeElement.GetAttribute(atHref);
+
+                if (valueOfHref == String.Empty) // There must be xpointer Attribute, if not, fatal error..
```
Comment (jakubmisek) [inline src/Peachpie.Library.XmlDom/XIncludeHelper.cs]: `string.IsNullOrEmpty(valueOfHref )`

File: src/Peachpie.Library.XmlDom/XIncludeHelper.cs [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+                }
+
+                // Resolving absolute and relative uri...
+                string uri = UriResolver(valueOfHref, Path.GetDirectoryName(absoluteUri));
+
+                // Resolving type of parsing.
+                string typeOfParse = includeElement.GetAttribute(atParse);
+                try
+                {
+                    if (typeOfParse == "text")
```
Comment (jakubmisek) [inline src/Peachpie.Library.XmlDom/XIncludeHelper.cs]: can it be case insensitive? `Text` ?

File: src/Peachpie.Library.XmlDom/XIncludeHelper.cs [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+                    {
+                        documents.Push(document);
+                        IncludeText(uri, includeElement);
+                    }
+                    else
+                    {                       
+                        if (valueOfXPoiner == String.Empty)
```
Comment (jakubmisek) [inline src/Peachpie.Library.XmlDom/XIncludeHelper.cs]: `string.IsNullOrEmpty`

File: src/Peachpie.Library.XmlDom/XIncludeHelper.cs [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+                                //return true;
+                            }
+                            else
+                            {
+                                documents.Push(document);
+                                references[absoluteUri] = uri;
+                                IncludeXml(uri, includeElement, null, null);
+                            }
+                        }
+                        else
```
Comment (jakubmisek) [inline src/Peachpie.Library.XmlDom/XIncludeHelper.cs]: duplicit code, just with a different value of `uri` 

line 224:
```C#
uri = string.IsNullOrEmpty(valueOfXPoiner) ? uri : (uri + valueOfXPoiner;);
// line 226 - 236
```

**Pull request merged**